Bug 1807072: Adds route-override-cni and whereabouts for secondary network routing and IPAM

dougbtv opened the pull request:
Updates YAML templates for init containers, primarily. Condenses
the binary copy script into a single place to be used across other
CNI plugin binary copies.

Likely dependent on https://github.com/openshift/release/pull/6815

juanluisvaladas commented:
/lgtm

juanluisvaladas commented:
/retest

danwinship commented on bindata/network/additional-networks/crd/001-crd.yaml:
```diff
@@ -13,3 +13,68 @@ spec:
+  scope: ""
+  validation:
+    openAPIV3Schema:
+      description: IPPool is the Schema for the ippools API
```
The documentation here is weird; it's there, but it's not enough to have any idea what any of this means unless you already know...

dougbtv replied:
Agreed, it's like using the word in a definition, I'll touch this up.

danwinship commented on bindata/network/additional-networks/crd/001-crd.yaml:
```diff
@@ -13,3 +13,68 @@ spec:
+  creationTimestamp: null
+  name: ippools.whereabouts.cni.k8s.io
+spec:
+  group: whereabouts.cni.k8s.io
```
If whereabouts isn't an official Kubernetes project then it shouldn't be using a `k8s.io` API group...

dougbtv replied:
Thanks for the eyes, this is an upstream oversight, it was intended to be cni.cncf.io.... I'll try to get it patched up...

danwinship commented on bindata/network/additional-networks/crd/001-crd.yaml:
```diff
@@ -13,3 +13,68 @@ spec:
+                - id
+                type: object
+              description: Allocations is the set of allocated IPs for the given range.
+                Its` indices are a direct mapping to the IP with the same index/offset
```
(stray backtick)

danwinship commented on bindata/network/multus/multus.yaml:
```diff
@@ -1,3 +1,82 @@
+  namespace: openshift-multus
+  annotations:
+    kubernetes.io/description: |
+      This is a script used to copy binaries in a FIPS compliant fashion
```
"copy binaries in a FIPS compliant fashion" makes it sound like you're using some cryptographically-secure version of `/usr/bin/cp`. This script really has nothing do to with FIPS per se, it's just that for FIPS reasons, we need separate RHEL7 and RHEL8 copies of CNI plugins, and this deals with that. I'd rename it all to be about installing CNI plugins rather than FIPS. And you can get rid of `DESTINATION_DIRECTORY`; it's always `/host/opt/cni/bin`.

Meanwhile, see also https://issues.redhat.com/browse/SDN-695. This script solves some of the problems there, though it still requires each daemonset to mount the right underlying files/dirs onto `/host/opt/cni/bin`, `/host/etc/os-release`, etc

dougbtv replied:
+1, yeah I'll change the language here to talk about cni binary copy based on host OS, makes more sense.

danwinship commented on bindata/network/multus/multus.yaml:
```diff
@@ -1,3 +1,82 @@
+data:
+  fipsbincopy.sh: |-
+    #!/bin/bash
+    set -x
```
No, don't do that (`set -x`) please. It just fills logs with junk

danwinship commented on bindata/network/multus/multus.yaml:
```diff
@@ -29,101 +108,204 @@ spec:
+        - name: DESTINATION_DIRECTORY
+          value: "/host/opt/cni/bin/"
+      - name: routeoverride-cni
+        image: {{.RouteOverrideImage}}
```
Do we really need to install this onto every node in every cluster? Can't we make this conditional on configuration somehow? (ie, it only gets deployed if you're using it). Similarly, does this really belong in the multus daemonset? It's really more "additional-networks" than "multus" isn't it?

Alternatively, perhaps we could copy all of the CNI binaries into a single image...

dougbtv replied:
Good considerations all. As for conditional copy, the only thing that really works this way is for DHCP CNI -- and in that case it's because it needs additional components (like the DHCP CNI running in daemon mode).

Previously, all the CNI binaries have been copied in these init containers for Multus... Well, the reference CNI plugins + the Multus binary, and now we add two more. I'll look at moving the CNI binary copies to be organized in the additional networks header...

I like the idea of condensing all of the CNI binaries into a single image, I think going forward that would likely be cleaner. However, there's a variety of upstream sources for these which adds some degree of complexity.

dougbtv commented:
New commit pushed addressing these issues:

* Updates language about CNI binary copy, also changes names to prefer "cni binary copy"
* Removes redundant `DESTINATION_DIRECTORY` from script
* Updates description of IPPool CRD (and removes stray backtick)

Still has route-override copied in the same location as the other CNI binaries in this daemonset.

Does not address the CRD namespace yet, I started an upstream conversation about it and looking into remedying that.

dougbtv commented:
Pushed another commit with updates to the CRD namespace

This depends on this PR: https://github.com/openshift/whereabouts-cni/pull/4

dougbtv commented:
/retest

dougbtv commented:
/retest

images should be public now

dougbtv commented:
/retest

Still failing with an ErrImagePull but, I did that as soon as I found they were public, trying again in case there's some kind of delay.

dougbtv commented:
/retest

Still seeing `errimagepull` but for the first time I see the OSBS builds are working, so wondering if other parts of the build system have changed.

dougbtv commented:
New commit pushed with the *proper* names, updated to `quay.io/openshift/origin-multus-whereabouts-ipam-cni:4.3` and `quay.io/openshift/origin-multus-route-override-cni:4.3` (had older names previously)

rcarrillocruz commented:
/retest

abhat commented on bindata/network/additional-networks/crd/001-crd.yaml:
```diff
@@ -26,3 +26,68 @@ spec:
+metadata:
+  creationTimestamp: null
+  name: ippools.whereabouts.cni.cncf.io
+spec:
```
add `preserveUnknownFields: true`

dougbtv replied:
Got it, thanks for the eyes Aniket!

dougbtv commented:
/test e2e-gcp-upgrade

dougbtv commented:
/retest

dougbtv commented:
/test e2e-gcp-upgrade

juanluisvaladas commented:
/retest

knobunc commented:
/bugzilla refresh

knobunc commented:
/lgtm

knobunc commented:
/retest
